### DISCUSSION TITLE
Developing and debugging local with Docker

### DISCUSSION THREAD

**Discussion by danielmcquillen** [cookiecutter/cookiecutter-django#5404] [General]
Hi all. I'm wading into the Docker-based approach to local dev, as it seems like the general consensus is the extra complexity pays off in the long run for deployments, consistency, etc.

However, I can't find much info on the latest best-practice to setting up VS Code to launch and debug the Django app (or Celery process). I see some similar questions in the posts on this forum seeking similar info, and have found some dated tutorials that only somewhat apply. 

Does anyone have pointers to a good resource that addresses debugging a Django project produced by Cookie cutter in a Docker configuration via VS Code? 

The official [docs page](https://cookiecutter-django.readthedocs.io/en/latest/developing-locally-docker.html) on this topic doesn't provide much info here, just a brief mention of ipdb (which I don't think is involved here).

I see that a .devcontainer folder is added during generation if one indicates they're using Docker and VS Code, (via https://github.com/cookiecutter/cookiecutter-django/pull/4198) but it's unclear to me how to leverage this in actual use. If I understand it correctly, these files help you set up a new VS Code within the running container itself? Is that the best way of debugging Django running in Docker?

It may be that running Django in docker locally isn't worth the extra complexity except for certain situations where you need exact parity with production environments. Would love to hear any thoughts on your own experience.

Thanks for any suggestions or pointers to extra info!

**Comment by foarsitter** [comment]
Using docker for local development solves the `works on my machine` problem and it even functions as documentation for your development environment.

I'm not a VSCode user myself but they provide docs about this topic that uses django as an example: https://code.visualstudio.com/docs/containers/debug-python can you give that a try?

**Comment by danielmcquillen** [comment]
Thanks @foarsitter. I'm trying to work through those docs and others...still various pieces I'm trying to synthesize. 

[This issue](https://github.com/cookiecutter/cookiecutter-django/issues/2580) mentions devcontainer.json, which seems to indicate best practice is to set up VS Code server inside a container? Not clear how one attaches to the running the Django app from this attached VS Code such that it respects breakpoints and can be debugged...

Tips from anyone who's got this working?

**Reply by quroom** [reply]
@foarsitter 
Me either.
So I just attached debugpy when run django app server.

Those Debug setting experience is not beautiful for developer, even pycharm([issue](https://intellij-support.jetbrains.com/hc/en-us/community/posts/360004951500-How-to-avoid-create-opt-project)).
I hope there is just one click setting or some script or config step in cookiecutter django.
